### ISSUE TITLE
Crashes in some environments: no default font found

### ISSUE THREAD

**Issue by nunotexbsd** [lapce/lapce#2748]
## Lapce Version

0.3.0 

## System information

- FreeBSD 15-CURRENT amd64
- Rust 1.7.3
- lapce compilation made without rust optimization

## Describe the bug
Crashes

## Additional information
```
env RUST_BACKTRACE=full lapce -w
thread 'main' panicked at /wrkdirs/usr/ports/editors/lapce/work/cosmic-text-5bfa288f19970a0b482ad75d5f460ad2c4e08900/src/buffer_line.rs:192:41:
no default font found
stack backtrace:
   0:     0x26a6681c863b - <unknown>
   1:     0x26a6681fff1c - <unknown>
   2:     0x26a6681bad0e - <unknown>
   3:     0x26a6681c8414 - <unknown>
   4:     0x26a6681afbe3 - <unknown>
   5:     0x26a6681af84f - <unknown>
   6:     0x26a6681b0e47 - <unknown>
   7:     0x26a6681c93b7 - <unknown>
   8:     0x26a6681c8846 - <unknown>
   9:     0x26a6681b0a72 - <unknown>
  10:     0x26a6682061e3 - <unknown>
  11:     0x26a6682124a3 - <unknown>
  12:     0x26a667f326ab - <unknown>
  13:     0x26a667f32758 - <unknown>
  14:     0x26a667f47667 - <unknown>
  15:     0x26a667f47e59 - <unknown>
  16:     0x26a667f473da - <unknown>
  17:     0x26a66660d0ec - <unknown>
  18:     0x26a665ffffa5 - <unknown>
  19:     0x26a66819bccb - <unknown>
  20:     0x26a66819f6f9 - <unknown>
  21:     0x26a665f247aa - <unknown>
  22:     0x26a66639cfee - <unknown>
  23:     0x26a665bd3f03 - <unknown>
  24:     0x26a665ca02d8 - <unknown>
  25:     0x26a665b92ad2 - <unknown>
  26:     0x26a665b510f9 - <unknown>
  27:     0x26a66775109f - <unknown>
  28:     0x26a6677db541 - <unknown>
  29:     0x26a667760e2e - <unknown>
  30:     0x26a665bb8f18 - <unknown>
  31:     0x26a665b2b7b3 - <unknown>
  32:     0x26a665b2b7c9 - <unknown>
  33:     0x26a6681afd57 - <unknown>
  34:     0x26a6681bf5bb - <unknown>
  35:     0x26a665b2b825 - <unknown>
  36:     0x26ae8da5be5a - __libc_start1
                               at /usr/src/lib/libc/csu/libc_start1.c:157:7
  37:     0x26a665b2b72d - <unknown>
```

FreeBSD issue: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=274852

Any hints?

**Comment by panekj** [comment]
> `no default font found`

**Comment by nunotexbsd** [comment]
@panekj 

Not used to rust backtrace info but indeed that line call my attention.

Should I update issue title with "no default font found"?

**Comment by nunotexbsd** [comment]
```
cosmic-text-5bfa288f19970a0b482ad75d5f460ad2c4e08900/src/shape.rs:    let font = font_iter.next().expect("no default font found");
cosmic-text-5bfa288f19970a0b482ad75d5f460ad2c4e08900/src/buffer_line.rs:            let font = font_iter.next().expect("no default font found");
```
Is lapce searching for a specific font?

**Comment by daniloegea** [comment]
```
```

**Comment by panekj** [comment]
This should be fixed in nightly already

**Comment by nunotexbsd** [comment]
@panekj 

Builded v0.3.0-7-g6766930d (latest nightly) and same error.

**Comment by Eliot00** [comment]
I'm trying to package the latest version on NiOS, I got an error:

```
error[E0658]: use of unstable library feature 'local_key_cell_methods'
   --> /build/cargo-vendor-dir/floem-0.1.0/src/inspector.rs:674:21
    |
674 |             RUNNING.set(false);
    |                     ^^^
    |
    = note: see issue #92122 <https://github.com/rust-lang/rust/issues/92122> for more information
    = help: add `#![feature(local_key_cell_methods)]` to the crate attributes to enable

   Compiling structdesc v0.1.0 (https://github.com/lapce/structdesc#a6a2b8ab)
   Compiling open v5.0.0
error[E0658]: use of unstable library feature 'local_key_cell_methods'
   --> /build/cargo-vendor-dir/floem-0.1.0/src/inspector.rs:706:17
    |
706 |     if !RUNNING.get() {
    |                 ^^^
    |
    = note: see issue #92122 <https://github.com/rust-lang/rust/issues/92122> for more information
    = help: add `#![feature(local_key_cell_methods)]` to the crate attributes to enable

error[E0658]: use of unstable library feature 'local_key_cell_methods'
   --> /build/cargo-vendor-dir/floem-0.1.0/src/inspector.rs:707:17
    |
707 |         RUNNING.set(true);
    |                 ^^^
    |
    = note: see issue #92122 <https://github.com/rust-lang/rust/issues/92122> for more information
    = help: add `#![feature(local_key_cell_methods)]` to the crate attributes to enable

   Compiling tracing-appender v0.2.0 (https://github.com/tokio-rs/tracing?rev=c14525e1610db88986f849d46bd3e9795878b012#c14525e1)
   Compiling nucleo v0.2.0
For more information about this error, try `rustc --explain E0658`.
error: could not compile `floem` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
error: builder for '/nix/store/k0pypy4klfcvzpln88il1674d7bgjb0b-lapce-unstable-2023-11-04.drv' failed with exit code 101;
       last 10 log lines:
       >     |                 ^^^
       >     |
       >     = note: see issue #92122 <https://github.com/rust-lang/rust/issues/92122> for more information
       >     = help: add `#![feature(local_key_cell_methods)]` to the crate attributes to enable
       >
       >    Compiling tracing-appender v0.2.0 (https://github.com/tokio-rs/tracing?rev=c14525e1610db88986f849d46bd3e9795878b012#c14525e1)
       >    Compiling nucleo v0.2.0
       > For more information about this error, try `rustc --explain E0658`.
       > error: could not compile `floem` (lib) due to 3 previous errors
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/k0pypy4klfcvzpln88il1674d7bgjb0b-lapce-unstable-2023-11-04.drv'.
```

**Comment by panekj** [comment]
You need to use latest stable rust

**Comment by nunotexbsd** [comment]
Still panics on 3.0.1:
```
env RUST_BACKTRACE=full lapce -w
2023-12-05T11:15:48.807658Z ERROR lapce_app::keypress: Failed to load from "/home/nunotex/.config/lapce-nightly/keymaps.toml": no keymaps
thread 'main' panicked at /wrkdirs/usr/ports/editors/lapce/work/cosmic-text-f7a20704d6ebbe8fb82d0bb579c37c53e7ae9747/src/buffer_line.rs:192:41:
no default font found
stack backtrace:
   0:     0x14bd4460b2cb - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::he796e85463f0e163
   1:     0x14bd44642dcc - core::fmt::write::h4263c96a220ebc07
   2:     0x14bd445fd99e - std::io::Write::write_fmt::hb2b0bfaf08bc4f10
   3:     0x14bd4460b0a4 - std::sys_common::backtrace::print::h9304fb7eae17b08b
   4:     0x14bd445f2873 - std::panicking::default_hook::{{closure}}::h00340db9279f127c
   5:     0x14bd445f24df - std::panicking::default_hook::h04c61a21bfe10d3c
   6:     0x14bd445f3ad7 - std::panicking::rust_panic_with_hook::h866a7df2ad4da1eb
   7:     0x14bd4460c047 - std::panicking::begin_panic_handler::{{closure}}::h93ad0edfea19b512
   8:     0x14bd4460b4d6 - std::sys_common::backtrace::__rust_end_short_backtrace::hea16e389464bca8b
   9:     0x14bd445f3702 - rust_begin_unwind
  10:     0x14bd44649093 - core::panicking::panic_fmt::h3af486d4a8a26501
  11:     0x14bd44655353 - core::option::expect_failed::hca4ca9da848ed557
  12:     0x14bd442c363b - cosmic_text::buffer_line::TextLayoutLine::shape::h0822020e2f6a77cf
  13:     0x14bd442c36e8 - cosmic_text::buffer_line::TextLayoutLine::layout::h9126429fec7dc236
  14:     0x14bd442f6ed7 - cosmic_text::buffer::TextLayout::shape_until::hf34d37504ddf05f2
  15:     0x14bd442f76c9 - cosmic_text::buffer::TextLayout::set_text::h4cae91458538aa49
  16:     0x14bd442f6c4a - cosmic_text::buffer::TextLayout::new::h610e6a92314e55d9
  17:     0x14bd427bbc4c - floem_reactive::signal::ReadSignal<T>::with_untracked::hc4aaf5e3ff009ee8
  18:     0x14bd42a6d84e - <floem_reactive::effect::Effect<T,F> as floem_reactive::effect::EffectTrait>::run::h44798a3151442d34
  19:     0x14bd445e2833 - floem_reactive::scope::with_scope::he3c575c5e73e3537
  20:     0x14bd445e360d - floem_reactive::effect::run_initial_effect::h42032fb281ca31f3
  21:     0x14bd428ae8e5 - floem_reactive::effect::create_effect::h0b84ecd04ddd650c
  22:     0x14bd428545ee - floem_reactive::scope::with_scope::hdfcf0ef36d9bb44c
  23:     0x14bd4232f8b8 - lapce_app::window_tab::WindowTabData::new::h6f75325390300b54
  24:     0x14bd4224d59f - lapce_app::window::WindowData::new::hde5b1c81fc68fd7d
  25:     0x14bd421a473a - lapce_app::app::AppData::app_view::he782e0f881d08491
  26:     0x14bd42155b69 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h445e95fdad6476cd
  27:     0x14bd43bab2df - floem_reactive::scope::with_scope::h5bb27963c003545e
  28:     0x14bd43b597a4 - floem::window_handle::WindowHandle::new::hdfaf94357c7d8a98
  29:     0x14bd43a9680b - floem::app_handle::ApplicationHandle::new_window::h4cf5f4689a063e5d
  30:     0x14bd421b82c8 - lapce_app::app::launch::hf99c049cbd73efbb
  31:     0x14bd42138233 - std::sys_common::backtrace::__rust_begin_short_backtrace::h46557505bb3d406c
  32:     0x14bd42138249 - std::rt::lang_start::{{closure}}::h5000938f48a2eff3
  33:     0x14bd445f29e7 - std::panicking::try::hc1d210acaca6f151
  34:     0x14bd4460224b - std::rt::lang_start_internal::hf93bb11da796dec0
  35:     0x14bd421382a5 - main
  36:     0x14c56b1ce14a - __libc_start1
                               at /usr/src/lib/libc/csu/libc_start1.c:157:7
  37:     0x14bd421381ad - _start
                               at /usr/src/lib/csu/amd64/crt1_s.S:83
```

**Comment by paulcarroty** [comment]
Got the same issue on Linux with 0.3.1.

**Comment by nunotexbsd** [comment]
@panekj

Looking at `lapce-0.3.1/cargo-crates/fontdb-0.16.0/src/lib.rs`:

```
    /// Attempts to load system fonts.
    ///
    /// Supports Windows, Linux and macOS.
    ///
    /// System fonts loading is a surprisingly complicated task,
    /// mostly unsolvable without interacting with system libraries.
    /// And since `fontdb` tries to be small and portable, this method
    /// will simply scan some predefined directories.
    /// Which means that fonts that are not in those directories must
    /// be added manually.
```
Maybe this block is missing the right call:

```
// Linux.
        #[cfg(all(unix, not(any(target_os = "macos", target_os = "android"))))]
        {
            #[cfg(feature = "fontconfig")]
            {
                self.load_fontconfig();
            }

            #[cfg(not(feature = "fontconfig"))]
            {
                self.load_fonts_dir("/usr/share/fonts/");
                self.load_fonts_dir("/usr/local/share/fonts/");

                if let Ok(ref home) = std::env::var("HOME") {
                    let home_path = std::path::Path::new(home);
                    self.load_fonts_dir(home_path.join(".fonts"));
                    self.load_fonts_dir(home_path.join(".local/share/fonts"));
                }
            }
        }

```

I wondering if something needed to make it to work on FreeBSD.

What do I need to open a PR at fondb.

**Comment by panekj** [comment]
> What do I need to open a PR at fondb.

There is nothing that needs to be done in fontdb, you need to have some kind of fonts installed on your system. If you use fontconfig, it will deal with that, if you don't then paths like below should contain fonts
```
                self.load_fonts_dir("/usr/share/fonts/");
                self.load_fonts_dir("/usr/local/share/fonts/");
```

my guess is that you don't have any fonts installed that would suffice for lapce

**Comment by nunotexbsd** [comment]
Ok, suposing that lapce will search for fonts at run time, it will find several directories with fonts inside at `"/usr/local/share/fonts/"` path.
The question is if it is looking for all fonts toguether in that path.

```
% ls /usr/local/share/fonts/
drwxr-xr-x  2 root wheel 1899 Dec 23 10:50 100dpi
drwxr-xr-x  2 root wheel 1899 Dec 23 10:50 75dpi
drwxr-xr-x  2 root wheel   27 Dec 23 10:49 OTF
drwxr-xr-x  2 root wheel   17 Dec 23 10:49 TTF
drwxr-xr-x  2 root wheel   61 Dec 23 10:50 Type1
drwxr-xr-x  2 root wheel   87 Dec 23 10:50 cyrillic
drwxr-xr-x  2 root wheel   26 Dec 23 10:49 dejavu
drwxr-xr-x  3 root wheel   37 Dec 23 10:49 encodings
drwxr-xr-x  2 root wheel  414 Dec 23 10:50 misc
```
```
% fc-list
/usr/local/share/fonts/75dpi/timB14.pcf.gz: Adobe Times:style=Bold
/usr/local/share/fonts/misc/clR4x6.pcf.gz: Schumacher Clean:style=Regular
/usr/local/share/fonts/100dpi/timI24.pcf.gz: Adobe Times:style=Italic
/usr/local/share/fonts/100dpi/ncenBI12.pcf.gz: Adobe New Century Schoolbook:style=Bold Italic
/usr/local/share/fonts/75dpi/lubB24.pcf.gz: B&H LucidaBright:style=Regular
/usr/local/share/fonts/100dpi/ncenR24-ISO8859-1.pcf.gz: Adobe New Century Schoolbook:style=Regular
/usr/local/share/fonts/100dpi/helvO14.pcf.gz: Adobe Helvetica:style=Oblique
/usr/local/share/fonts/75dpi/helvB08-ISO8859-1.pcf.gz: Adobe Helvetica:style=Bold
/usr/local/share/fonts/75dpi/helvB10-ISO8859-1.pcf.gz: Adobe Helvetica:style=Bold
/usr/local/share/fonts/75dpi/helvB14-ISO8859-1.pcf.gz: Adobe Helvetica:style=Bold
/usr/local/share/fonts/75dpi/luIS12.pcf.gz: B&H Lucida:style=Sans Italic
/usr/local/share/fonts/75dpi/ncenR08.pcf.gz: Adobe New Century Schoolbook:style=Regular
/usr/local/share/fonts/75dpi/ncenR10.pcf.gz: Adobe New Century Schoolbook:style=Regular
/usr/local/share/fonts/Type1/c0649bt_.pfb: Bitstream Charter:style=Italic
/usr/local/share/fonts/100dpi/timR12.pcf.gz: Adobe Times:style=Regular
/usr/local/share/fonts/75dpi/luRS12.pcf.gz: B&H Lucida:style=Sans
/usr/local/share/fonts/75dpi/lutRS19-ISO8859-1.pcf.gz: B&H LucidaTypewriter:style=Sans
/usr/local/share/fonts/100dpi/luRS19-ISO8859-1.pcf.gz: B&H Lucida:style=Sans
/usr/local/share/fonts/misc/clR8x12.pcf.gz: Schumacher Clean:style=Regular
/usr/local/share/fonts/100dpi/UTRG__12-ISO8859-1.pcf.gz: Adobe Utopia:style=Regular
/usr/local/share/fonts/100dpi/helvB08.pcf.gz: Adobe Helvetica:style=Bold
/usr/local/share/fonts/100dpi/helvB10.pcf.gz: Adobe Helvetica:style=Bold
/usr/local/share/fonts/100dpi/lubB18.pcf.gz: B&H LucidaBright:style=Regular
/usr/local/share/fonts/75dpi/charR12.pcf.gz: Bitstream Charter:style=Regular
/usr/local/share/fonts/dejavu/DejaVuSerifCondensed.ttf: DejaVu Serif,DejaVu Serif Condensed:style=Condensed,Book
/usr/local/share/fonts/100dpi/lubI24-ISO8859-1.pcf.gz: B&H LucidaBright:style=Italic
/usr/local/share/fonts/100dpi/luIS19.pcf.gz: B&H Lucida:style=Sans Italic
/usr/local/share/fonts/75dpi/helvBO12-ISO8859-1.pcf.gz: Adobe Helvetica:style=Bold Oblique
/usr/local/share/fonts/100dpi/lutRS19-ISO8859-1.pcf.gz: B&H LucidaTypewriter:style=Sans
/usr/local/share/fonts/75dpi/timR18.pcf.gz: Adobe Times:style=Regular
/usr/local/share/fonts/75dpi/timI12-ISO8859-1.pcf.gz: Adobe Times:style=Italic
/usr/local/share/fonts/100dpi/luBS24-ISO8859-1.pcf.gz: B&H Lucida:style=Sans Bold
/usr/local/share/fonts/75dpi/timR24-ISO8859-1.pcf.gz: Adobe Times:style=Regular
/usr/local/share/fonts/TTF/luxisbi.ttf: Luxi Sans:style=Bold Oblique
<snip etc>
```

**Comment by paulcarroty** [comment]
> my guess is that you don't have any fonts installed that would suffice for lapce

any examples of?

**Comment by volkertb** [comment]
It's not just an issue in FreeBSD. I'm getting the same `no default font found` error with lapce `v0.3.1` when I tried to start the Linux executable (from [Lapce-linux.tar.gz](https://github.com/lapce/lapce/releases/download/v0.3.1/Lapce-linux.tar.gz)) in Clear Linux OS version 40580:

```
$ RUST_BACKTRACE=full lapce -w
2023-12-31T10:26:27.257338Z ERROR lapce_app::keypress: Failed to load from "[HOME FOLDER OMITTED]/.config/lapce-stable/keymaps.toml": no keymaps
thread 'main' panicked at /github/home/.cargo/git/checkouts/cosmic-text-fff8c4a78b1e732f/f7a2070/src/buffer_line.rs:192:41:
no default font found
stack backtrace:
   0:     0x55b7d93b72cf - std::backtrace_rs::backtrace::libunwind::trace::h67a838aed1f4d6ec
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55b7d93b72cf - std::backtrace_rs::backtrace::trace_unsynchronized::h1d1786bb1962baf8
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55b7d93b72cf - std::sys_common::backtrace::_print_fmt::h5a0b1f807a002d23
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x55b7d93b72cf - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf84ab6ad0b91784c
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x55b7d7e4768c - core::fmt::rt::Argument::fmt::h28f463bd1fdabed5
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/fmt/rt.rs:138:9
   5:     0x55b7d7e4768c - core::fmt::write::ha37c23b175e921b3
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/fmt/mod.rs:1114:21
   6:     0x55b7d93809fd - std::io::Write::write_fmt::haa1b000741bcbbe1
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/io/mod.rs:1763:15
   7:     0x55b7d93b996e - std::sys_common::backtrace::_print::h1ff1030b04dfb157
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x55b7d93b996e - std::sys_common::backtrace::print::hb982056c6f29541c
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x55b7d93b9553 - std::panicking::default_hook::{{closure}}::h11f92f82c62fbd68
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:272:22
  10:     0x55b7d93ba519 - std::panicking::default_hook::hb8810fe276772c66
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:292:9
  11:     0x55b7d93ba519 - std::panicking::rust_panic_with_hook::hd2f0efd2fec86cb0
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:731:13
  12:     0x55b7d93ba04c - std::panicking::begin_panic_handler::{{closure}}::h3651b7fc4f61d784
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:609:13
  13:     0x55b7d93b9fa6 - std::sys_common::backtrace::__rust_end_short_backtrace::hbc468e4b98c7ae04
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/sys_common/backtrace.rs:170:18
  14:     0x55b7d93b9f91 - rust_begin_unwind
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:597:5
  15:     0x55b7d7c6ecb4 - core::panicking::panic_fmt::h979245e2fdb2fabd
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:72:14
  16:     0x55b7d7c6f292 - core::panicking::panic_display::h9b355c58fd35af38
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:168:5
  17:     0x55b7d7c6f292 - core::panicking::panic_str::h187a5146d72e7d2f
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:152:5
  18:     0x55b7d7c6f292 - core::option::expect_failed::h7cdfa49208a82a89
                               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/option.rs:1988:5
  19:     0x55b7d7e68fa5 - cosmic_text::buffer_line::TextLayoutLine::layout::h19821fe909a45848
  20:     0x55b7d7e607a6 - cosmic_text::buffer::TextLayout::shape_until::h47a6915dcaf6abe0
  21:     0x55b7d7e6013c - cosmic_text::buffer::TextLayout::set_text::h62e24f6d61f9feb2
  22:     0x55b7d7e5e396 - cosmic_text::buffer::TextLayout::new::h68aefb1338c88f2c
  23:     0x55b7d89c978c - <floem_reactive::effect::Effect<T,F> as floem_reactive::effect::EffectTrait>::run::hd4c3d02be736e833
  24:     0x55b7d81686a5 - floem_reactive::effect::run_initial_effect::h9013e2c0c4bcd1cd
  25:     0x55b7d893f0d3 - lapce_app::window_tab::WindowTabData::new::h8825f545ba8d008a
  26:     0x55b7d8397df8 - lapce_app::app::AppData::app_view::h359901ecee10803f
  27:     0x55b7d8bc5deb - core::ops::function::FnOnce::call_once{{vtable.shim}}::h35698ecc62d5408d
  28:     0x55b7d80b1780 - floem::window_handle::WindowHandle::new::hcc330a3260bf3529
  29:     0x55b7d809ecac - floem::app_handle::ApplicationHandle::new_window::h3a552fbbb49b754c
  30:     0x55b7d8b8e95b - lapce_app::app::launch::h46fc00ea5799c667
  31:     0x55b7d7d56623 - std::sys_common::backtrace::__rust_begin_short_backtrace::hbd23bd1a12cd408d
  32:     0x55b7d7d56a95 - main
  33:     0x55b7d7227ac7 - __libc_start_call_main
                               at /builddir/build/BUILD/glibc-2.38/csu/../sysdeps/nptl/libc_start_call_main.h:58:16
  34:     0x55b7d7227b85 - __libc_start_main_impl
                               at /builddir/build/BUILD/glibc-2.38/csu/../csu/libc-start.c:360:3
  35:     0x55b7d7d5610a - _start
  36:                0x0 - <unknown>
```

As you can see in the output, there are two errors: the  `no keymaps` error, which appeared to be non-fatal, and the `no default font found` error, which apparently caused the panic.

Also, issues #2752, #2757 and #2837 also make it likely that this is not a FreeBSD-specific issue, but more likely a dependency in the current version of lapce that only happens to work out of the box on some Linux distros, but not on others.

I also just tried this with the latest nightly `Lapce-linux.tar.gz` as of 2023-12-31(with SHA-256 checksum `7862ef3baef157ef6e81d01f14c7cff43e7f5214322e38c58ffbc681979ec643` and that build still has the same issue.

Looks like we're waiting on a fixed version of the `fontdb` crate to be released with a fix for this, right, @nunotexbsd?

Maybe someone can rename the title of this ticket to `Crashes in some environments: no default font found`?

**Comment by nunotexbsd** [comment]
@panekj 

Finally get it to run.

fontdb:
```
--- cargo-crates/fontdb-0.16.0/src/lib.rs.orig  2006-07-24 01:21:28 UTC
+++ cargo-crates/fontdb-0.16.0/src/lib.rs
@@ -444,8 +444,10 @@ impl Database {

             if read_global {
                 let _ = fontconfig.merge_config(Path::new("/etc/fonts/local.conf"));
+                let _ = fontconfig.merge_config(Path::new("/usr/local/etc/fonts/local.conf"));
             }
             let _ = fontconfig.merge_config(Path::new("/etc/fonts/fonts.conf"));
+            let _ = fontconfig.merge_config(Path::new("/usr/local/etc/fonts/fonts.conf"));
         }

         for fontconfig_parser::Alias {
```
and
cosmic-text:
```
--- ../cosmic-text-f7a20704d6ebbe8fb82d0bb579c37c53e7ae9747/src/font/fallback/mod.rs.orig       2023-12-31 09:09:22.570192000 +0000
+++ ../cosmic-text-f7a20704d6ebbe8fb82d0bb579c37c53e7ae9747/src/font/fallback/mod.rs    2023-12-31 09:09:48.358365000 +0000
@@ -11,7 +11,7 @@
 use self::platform::*;

 #[cfg(not(any(target_os = "linux", target_os = "macos", target_os = "windows",)))]
-#[path = "other.rs"]
+#[path = "unix.rs"]
 mod platform;

 #[cfg(target_os = "macos")]
```
Fixed by using "unix.rs" instead of "other.rs" for platforms that do not match that list.
Any sugestion of including "freebsd" target?

**Comment by volkertb** [comment]
There is also [this suggestion](https://github.com/pop-os/cosmic-text/issues/167#issuecomment-1676705605) to fix this error in NixOS, at least for `cosmic-text`:

> Adding features = ["fontconfig"] to fontdb in the cargo.toml fixes the doctest.

**Comment by nunotexbsd** [comment]
> There is also [this suggestion](https://github.com/pop-os/cosmic-text/issues/167#issuecomment-1676705605) to fix this error in NixOS, at least for `cosmic-text`:
> 
> > Adding features = ["fontconfig"] to fontdb in the cargo.toml fixes the doctest.

fontdb shiped with 0.3.1 does it by default.
```
fontdb-0.16.0/Cargo.toml:

[features]
default = [
    "std",
    "fs",
    "memmap",
    "fontconfig",
]
fontconfig = [
    "fontconfig-parser",
    "fs",
]
```

**Comment by volkertb** [comment]
@nunotexbsd I just searched for both `local.conf` and `fonts.conf` in Clear Linux OS, and in this distros, the paths appear to be different still:

- /usr/share/defaults/fonts/conf.d/51-local.conf
- /usr/share/fontconfig/conf.avail/51-local.conf
- /usr/share/defaults/fonts/fonts.conf

It feels like instead of making assumptions where each different distro or Unix-derative/POSIX OS places these files, deferring to `fontconfig`, as appears to be the solution that @jalil-salame proposed in the corresponding `cosmic-text` issue, seems like the proper way to go.

**Comment by volkertb** [comment]
> > There is also [this suggestion](https://github.com/pop-os/cosmic-text/issues/167#issuecomment-1676705605) to fix this error in NixOS, at least for `cosmic-text`:
> > > Adding features = ["fontconfig"] to fontdb in the cargo.toml fixes the doctest.
> 
> fontdb shiped with 0.3.1 does it by default.
> 
> ```
> fontdb-0.16.0/Cargo.toml:
> 
> [features]
> default = [
>     "std",
>     "fs",
>     "memmap",
>     "fontconfig",
> ]
> fontconfig = [
>     "fontconfig-parser",
>     "fs",
> ]
> ```

Aw, crap. So that doesn't fix it. Do we really need to start manually adding different paths for each different distro or OS, then? :confused:

**Comment by nunotexbsd** [comment]
> > > There is also [this suggestion](https://github.com/pop-os/cosmic-text/issues/167#issuecomment-1676705605) to fix this error in NixOS, at least for `cosmic-text`:
> > > > Adding features = ["fontconfig"] to fontdb in the cargo.toml fixes the doctest.
> > 
> > 
> > fontdb shiped with 0.3.1 does it by default.
> > ```
> > fontdb-0.16.0/Cargo.toml:
> > 
> > [features]
> > default = [
> >     "std",
> >     "fs",
> >     "memmap",
> >     "fontconfig",
> > ]
> > fontconfig = [
> >     "fontconfig-parser",
> >     "fs",
> > ]
> > ```
> 
> Aw, crap. So that doesn't fix it. Do we really need to start manually adding different paths for each different distro or OS, then? Surely there has got to be a multi-platform method to detect installed fonts, right? 😕

That point was added to:
https://github.com/RazrFalcon/fontdb/issues/61

**Comment by volkertb** [comment]
@nunotexbsd I just tried creating this symbolic link, and lapce still failed with `no default font found` :

```
sudo mkdir -p /usr/local/etc/fonts
ln -s /usr/share/defaults/fonts/fonts.conf /usr/local/etc/fonts/fonts.conf
lapce
```

Then I tried creating this symbolic link instead (after deleting the above non-working one):

```
sudo mkdir /etc/fonts
sudo ln -s /usr/share/defaults/fonts/fonts.conf /etc/fonts/fonts.conf
lapce
```

And that one worked! (At least with the nightly build of lapce.)

**Comment by volkertb** [comment]
Although that workaround works for me, when I run `lapce -w`, I still get the `no keymaps` error, since the file `$HOME/.config/lapce-nightly/keymaps.toml` doesn't exist yet. Which makes sense, since I'm starting the application for the first time. Maybe don't log this as an error? (Probably best to open a separate ticket for that?)

**Comment by panekj** [comment]
Keymaps message has correct log level, please do not open any issue about that or change it

**Comment by volkertb** [comment]
So I assume that the idea is that future versions of lapce will eventually automatically create a default keymaps file in `$HOME/.config` on startup when one doesn't exist yet?

**Comment by panekj** [comment]
No, but it's offtopic for this issue

**Comment by nunotexbsd** [comment]
@panekj

I think we can close this PR.

Run is fixed by using "unix.rs" and not "other.rs" (since freebsd isn't listed) in cosmic-text:
```
--- ../cosmic-text-f7a20704d6ebbe8fb82d0bb579c37c53e7ae9747/src/font/fallback/mod.rs.orig       2023-12-31 09:09:22.570192000 +0000
+++ ../cosmic-text-f7a20704d6ebbe8fb82d0bb579c37c53e7ae9747/src/font/fallback/mod.rs    2023-12-31 09:09:48.358365000 +0000
@@ -11,7 +11,7 @@
 use self::platform::*;

 #[cfg(not(any(target_os = "linux", target_os = "macos", target_os = "windows",)))]
-#[path = "other.rs"]
+#[path = "unix.rs"]
 mod platform;

 #[cfg(target_os = "macos")]
```

So, this is a cosmic-text issue to be discussed with upstream about adding "freebsd" support to it.

Any sugestions to open an PR?

**Comment by VPelmeSHko** [comment]
Got the same issue on fresh Archlinux install with Lapce 0.3.1 (from pacman) and Lapce [nightly](https://github.com/lapce/lapce/commit/2c08177d9b937e71bf4976c112372c8dd9a46116)
```
RUST_BACKTRACE=full ./lapce -w                                                                                                                                101 
2024-01-25T13:41:59.204834Z ERROR lapce_app::keypress: Failed to load from "/home/user/.config/lapce-nightly/keymaps.toml": no keymaps
thread 'main' panicked at /github/home/.cargo/git/checkouts/cosmic-text-fff8c4a78b1e732f/fdc5165/src/buffer_line.rs:192:41:
no default font found
stack backtrace:
   0:     0x5da4026878ac - std::backtrace_rs::backtrace::libunwind::trace::ha637c64ce894333a
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x5da4026878ac - std::backtrace_rs::backtrace::trace_unsynchronized::h47f62dea28e0c88d
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x5da4026878ac - std::sys_common::backtrace::_print_fmt::h9eef0abe20ede486
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x5da4026878ac - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hed7f999df88cc644
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x5da401071700 - core::fmt::rt::Argument::fmt::h1539a9308b8d058d
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/fmt/rt.rs:142:9
   5:     0x5da401071700 - core::fmt::write::h3a39390d8560d9c9
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/fmt/mod.rs:1120:17
   6:     0x5da40264fa42 - std::io::Write::write_fmt::h5fc9997dfe05f882
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/io/mod.rs:1762:15
   7:     0x5da40268a67e - std::sys_common::backtrace::_print::h894006fb5c6f3d45
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x5da40268a67e - std::sys_common::backtrace::print::h23a2d212c6fff936
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x5da402689df0 - std::panicking::default_hook::{{closure}}::h8a1d2ee00185001a
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:272:22
  10:     0x5da40268afbb - std::panicking::default_hook::h6038f2eba384e475
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:292:9
  11:     0x5da40268afbb - std::panicking::rust_panic_with_hook::h2b5517d590cab22e
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:779:13
  12:     0x5da40268a9dc - std::panicking::begin_panic_handler::{{closure}}::h233112c06e0ef43e
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:657:13
  13:     0x5da40268a936 - std::sys_common::backtrace::__rust_end_short_backtrace::h6e893f24d7ebbff8
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:170:18
  14:     0x5da40268a921 - rust_begin_unwind
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
  15:     0x5da400e7f8f4 - core::panicking::panic_fmt::hbf0e066aabfa482c
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
  16:     0x5da400e7fed2 - core::panicking::panic_display::h68e6012a1c287ca5
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:178:5
  17:     0x5da400e7fed2 - core::panicking::panic_str::h1ca8e3b0d2d70f80
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:152:5
  18:     0x5da400e7fed2 - core::option::expect_failed::h1ae4bac6b7606972
                               at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/option.rs:1985:5
  19:     0x5da40109299a - cosmic_text::buffer_line::TextLayoutLine::shape::hff433d855890b0a6
  20:     0x5da40108a731 - cosmic_text::buffer_line::TextLayoutLine::layout::h65afe94e476ae33d
  21:     0x5da401089f26 - cosmic_text::buffer::TextLayout::shape_until::h10b3bebd9ec97e30
  22:     0x5da401089809 - cosmic_text::buffer::TextLayout::set_text::h593121bec6e31abf
  23:     0x5da401087aa6 - cosmic_text::buffer::TextLayout::new::h5880538358412a85
  24:     0x5da401c5aa3c - <floem_reactive::effect::Effect<T,F> as floem_reactive::effect::EffectTrait>::run::hc49c67dab317611a
  25:     0x5da4013a3f57 - floem_reactive::effect::run_initial_effect::h32066d0889dee4ff
  26:     0x5da401bc940a - lapce_app::window_tab::WindowTabData::new::h7fdba916dd66f15b
  27:     0x5da4015d64a9 - lapce_app::app::AppData::app_view::h7d1565def926e2e5
  28:     0x5da401e5a51b - core::ops::function::FnOnce::call_once{{vtable.shim}}::h222c242c84ff1a7d
  29:     0x5da401314ecd - floem::window_handle::WindowHandle::new::hf9ee23d2f21bcf97
  30:     0x5da4013042df - floem::app_handle::ApplicationHandle::new_window::ha0df48420b6c3459
  31:     0x5da401e1acf8 - lapce_app::app::launch::h2ba10bcd8c39b680
  32:     0x5da400f7b9d3 - std::sys_common::backtrace::__rust_begin_short_backtrace::h27eca8bb5fb136b5
  33:     0x5da400f7bde9 - std::rt::lang_start::h9a8319cf1c065fe0
  34:     0x72415fb58cd0 - <unknown>
  35:     0x72415fb58d8a - __libc_start_main
  36:     0x5da400f7b4ba - _start
  37:                0x0 - <unknown>
```

```
$ ls -1 /etc/fonts
conf.d
fonts.conf
```

```
$ fc-match
DroidSans.ttf: "Droid Sans" "Regular"
```

**Comment by nunotexbsd** [comment]
@VPelmeSHko 

Try patching:
```
--- cargo-crates/fontdb-0.16.0/src/lib.rs.orig	2006-07-24 01:21:28 UTC
+++ cargo-crates/fontdb-0.16.0/src/lib.rs
@@ -444,8 +444,10 @@ impl Database {
 
             if read_global {
                 let _ = fontconfig.merge_config(Path::new("/etc/fonts/local.conf"));
+                let _ = fontconfig.merge_config(Path::new("/usr/local/etc/fonts/local.conf"));
             }
             let _ = fontconfig.merge_config(Path::new("/etc/fonts/fonts.conf"));
+            let _ = fontconfig.merge_config(Path::new("/usr/local/etc/fonts/fonts.conf"));
         }
 
         for fontconfig_parser::Alias {
```
to your font path

**Comment by nunotexbsd** [comment]
Oops, crate already hardcoded to your fonts path ...

**Comment by VPelmeSHko** [comment]
Already tried to add my Droid font in $HOME/.config/lapce-stable/settings.toml, but no luck, got same error
```
[editor]
font-family = "Droid Sans"
font-size = 13
code-lens-font-size = 2
line-height = 1.5
smart-tab = true

[terminal]
font-family = "Droid Sans"
font-size = 12
line-height = 0

[ui]
font-family = "Droid Sans"
font-size = 13
```
